File: backend/app/services/agent_registry.py
"""Agent registration and session management."""

import os
import uuid
from datetime import UTC, datetime

from loguru import logger
from sqlalchemy import func, select, update

from backend.app.db import async_session
from backend.app.models.agent import Agent
from backend.app.models.channel import Channel
from backend.app.models.channel_member import ChannelMember
from backend.app.models.feature import FeatureRequest, FeatureVote
from backend.app.models.session import Session
from backend.app.models.user import User
from backend.app.services.agent_discovery import auto_discover
from backend.app.services.broadcaster import (
    agent_status_event,
    broadcast_event,
    channel_created_event,
    feature_update_event,
)
from backend.app.services.name_generator import generate_unique_name
from backend.app.services.prompt_engine import prompt_engine


async def _derive_project_name(project_path: str) -> str:
    """Derive project name from path (git repo name or folder basename)."""
    import asyncio

    try:
        proc = await asyncio.create_subprocess_exec(
            "git",
            "-C",
            project_path,
            "rev-parse",
            "--show-toplevel",
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=5)
        if proc.returncode == 0:
            return os.path.basename(stdout.decode().strip())
    except (FileNotFoundError, TimeoutError):
        pass
    return os.path.basename(os.path.normpath(project_path))


def _make_channel_name(project_name: str) -> str:
    """Create a channel name from project name."""
    clean = project_name.lower().replace(" ", "-").replace("_", "-")
    return f"#project-{clean}"


async def register_or_connect_agent(
    session_id: str,
    project_path: str,
    agent_name: str | None = None,
    server_url: str | None = None,
) -> dict:
    """Register a new agent or reconnect an existing one.

    This is the single entry point for agent login. The ``session_id`` is
    required — it's the agent's login credential.

    - If ``agent_name`` is provided and exists in the DB, reconnect that
      identity with the new session_id (like the old ``connect()``).
    - Otherwise, create a fresh agent with a new unique name.

    The ``agent_type`` is always set to ``"opencode"`` — all agents run
    inside OpenCode and the type is determined server-side.
    """
    logger.info(
        "[REG] register_or_connect_agent: session_id={} agent_name={} project_path={}",
        session_id,
        agent_name,
        project_path,
    )
    agent_type = "opencode"

    # Auto-discover server_url if not provided
    if not server_url:
        try:
            discovery = await auto_discover(project_path)
            if discovery:
                server_url = discovery.get("server_url")
                logger.info("Auto-discovered server_url: %s", server_url)
        except Exception:
            logger.exception("Auto-discovery failed during registration")

    # --- Reconnect path: agent_name provided and exists ---
    if agent_name:
        async with async_session() as db:
            result = await db.execute(select(Agent).where(Agent.agent_name == agent_name))
            agent = result.scalar_one_or_none()
            if agent:
                logger.info("[REG] Reconnecting existing agent '{}'", agent_name)
                return await _reconnect_agent(db, agent, session_id, server_url, project_path)
        logger.info("[REG] Agent name '{}' not found — creating new agent", agent_name)
        # Agent name was provided but doesn't exist — fall through to create new

    # --- New registration path ---
    return await _create_new_agent(
        session_id=session_id,
        project_path=project_path,
        agent_type=agent_type,
        server_url=server_url,
    )


async def _reconnect_agent(
    db,
    agent: Agent,
    session_id: str,
    server_url: str | None,
    project_path: str,
) -> dict:
    """Reconnect an existing agent with a new session_id."""
    logger.info(
        "[REG] _reconnect_agent: name={} session_id={} server_url={}",
        agent.agent_name,
        session_id,
        server_url,
    )
    # Update invocation fields
    if server_url:
        agent.server_url = server_url
    agent.provider_session_id = session_id
    agent.agent_type = "opencode"  # Normalize — all agents are opencode now

    # Update project_path if provided (agent may have moved)
    if project_path:
        agent.project_path = project_path
        agent.project_name = await _derive_project_name(project_path)

    agent.status = "online"
    await db.commit()

    # Broadcast agent online status
    await broadcast_event(
        agent_status_event(
            agent_name=agent.agent_name,
            status="online",
            agent_type=agent.agent_type,
            project_name=agent.project_name,
        )
    )

    # Look up human operator for prompt context
    human_result = await db.execute(select(User).where(User.type == "human"))
    human = human_result.scalar_one_or_none()

    channel_name = _make_channel_name(agent.project_name)
    master_prompt = prompt_engine.render_master_prompt(
        agent_name=agent.agent_name,
        agent_type=agent.agent_type,
        project_name=agent.project_name,
        project_channel=channel_name,
        operator_name=human.name if human else "",
        operator_display_name=human.display_name or "" if human else "",
        operator_about=human.about or "" if human else "",
        operator_instructions=human.agent_instructions or "" if human else "",
    )
    inject_prompt = prompt_engine.render_registration_rules(
        agent_name=agent.agent_name,
        project_channel=channel_name,
    )

    # Build profile reminder so agent knows what they previously set
    profile: dict[str, str] = {}
    if agent.description:
        profile["description"] = agent.description
    if agent.personality:
        profile["personality"] = agent.personality
    if agent.current_task:
        profile["current_task"] = agent.current_task
    if agent.gender:
        profile["gender"] = agent.gender

    logger.info("[REG] Reconnect complete: agent={} channel={}", agent.agent_name, channel_name)
    return {
        "status": "connected",
        "agent_name": agent.agent_name,
        "project_channel": channel_name,
        "master_prompt": master_prompt,
        "inject_prompt": inject_prompt,
        "profile": profile if profile else None,
    }


async def _create_new_agent(
    session_id: str,
    project_path: str,
    agent_type: str,
    server_url: str | None,
) -> dict:
    """Create a brand new agent identity."""
    logger.info("[REG] Creating new agent: session_id={} project_path={}", session_id, project_path)
    now = datetime.now(UTC).isoformat()
    project_name = await _derive_project_name(project_path)
    channel_name = _make_channel_name(project_name)

    async with async_session() as db:
        # Generate a unique quirky name
        attempt = 0
        while True:
            agent_name = generate_unique_name(project_name, agent_type, attempt)
            collision = await db.execute(select(Agent).where(Agent.agent_name == agent_name))
            if collision.scalar_one_or_none() is None:
                break
            attempt += 1

        # Create user
        user_id = str(uuid.uuid4())
        user = User(id=user_id, name=agent_name, type="agent", created_at=now)
        db.add(user)

        # Create agent
        agent = Agent(
            id=user_id,
            agent_name=agent_name,
            agent_type=agent_type,
            project_path=project_path,
            project_name=project_name,
            status="online",
            server_url=server_url,
            provider_session_id=session_id,
        )
        db.add(agent)

        # Ensure project channel exists
        ch_result = await db.execute(select(Channel).where(Channel.name == channel_name))
        channel = ch_result.scalar_one_or_none()
        created_channel = False
        if not channel:
            channel = Channel(
                id=str(uuid.uuid4()),
                name=channel_name,
                type="project",
                project_path=project_path,
                created_by=user_id,
                created_at=now,
            )
            db.add(channel)
            await db.flush()
            created_channel = True

        # Add agent to project channel and #general
        db.add(ChannelMember(channel_id=channel.id, user_id=user_id, joined_at=now))

        general_result = await db.execute(select(Channel).where(Channel.name == "#general"))
        general = general_result.scalar_one_or_none()
        if general:
            db.add(ChannelMember(channel_id=general.id, user_id=user_id, joined_at=now))

        await db.commit()

        # Broadcast agent online status
        await broadcast_event(
            agent_status_event(
                agent_name=agent_name,
                status="online",
                agent_type=agent_type,
                project_name=project_name,
            )
        )

        # Broadcast new channel creation if we created one
        if created_channel:
            await broadcast_event(
                channel_created_event(
                    channel_id=channel.id,
                    channel_name=channel.name,
                    channel_type="project",
                    project_path=project_path,
                )
            )

        # Look up human operator for prompt context
        human_result = await db.execute(select(User).where(User.type == "human"))
        human = human_result.scalar_one_or_none()

        # Render prompts
        master_prompt = prompt_engine.render_master_prompt(
            agent_name=agent_name,
            agent_type=agent_type,
            project_name=project_name,
            project_channel=channel_name,
            operator_name=human.name if human else "",
            operator_display_name=human.display_name or "" if human else "",
            operator_about=human.about or "" if human else "",
            operator_instructions=human.agent_instructions or "" if human else "",
        )
        inject_prompt = prompt_engine.render_registration_rules(
            agent_name=agent_name,
            project_channel=channel_name,
        )

        logger.info(
            "[REG] New agent created: name={} channel={} type={}",
            agent_name,
            channel_name,
            agent_type,
        )
        return {
            "agent_name": agent_name,
            "master_prompt": master_prompt,
            "project_channel": channel_name,
            "inject_prompt": inject_prompt,
        }


async def disconnect_agent(agent_name: str) -> dict:
    """Mark agent as offline and end active session."""
    now = datetime.now(UTC).isoformat()

    async with async_session() as db:
        result = await db.execute(select(Agent).where(Agent.agent_name == agent_name))
        agent = result.scalar_one_or_none()
        if not agent:
            return {"error": f"Agent '{agent_name}' not found."}

        await db.execute(
            update(Session)
            .where(Session.agent_id == agent.id, Session.is_active == 1)
            .values(is_active=0, ended_at=now)
        )
        agent.status = "offline"
        await db.commit()

        # Broadcast agent offline status
        await broadcast_event(
            agent_status_event(
                agent_name=agent_name,
                status="offline",
                agent_type=agent.agent_type,
                project_name=agent.project_name,
            )
        )

        return {"status": "disconnected", "agent_name": agent_name}


async def heartbeat_agent(agent_name: str) -> dict:
    """Update agent's last heartbeat."""
    now = datetime.now(UTC).isoformat()

    async with async_session() as db:
        result = await db.execute(select(Agent).where(Agent.agent_name == agent_name))
        agent = result.scalar_one_or_none()
        if not agent:
            return {"error": f"Agent '{agent_name}' not found."}

        await db.execute(
            update(Session)
            .where(Session.agent_id == agent.id, Session.is_active == 1)
            .values(last_heartbeat=now)
        )
        await db.commit()

        return {"status": "ok", "agent_name": agent_name}


async def list_all_agents() -> list[dict]:
    """List all registered agents with profile info."""
    async with async_session() as db:
        result = await db.execute(select(Agent).order_by(Agent.agent_name))
        agents = []
        for a in result.scalars().all():
            entry: dict = {
                "name": a.agent_name,
                "type": a.agent_type,
                "project": a.project_name,
                "status": a.status,
            }
            if a.description:
                entry["description"] = a.description
            if a.personality:
                entry["personality"] = a.personality
            if a.current_task:
                entry["current_task"] = a.current_task
            if a.gender:
                entry["gender"] = a.gender
            agents.append(entry)
        return agents


async def agent_vote_feature(agent_name: str, feature_id: str, vote: int) -> dict:
    """Cast a vote on a feature request as an agent."""
    if vote not in (1, -1):
        return {"error": "Vote must be +1 or -1"}

    async with async_session() as db:
        agent_result = await db.execute(select(Agent).where(Agent.agent_name == agent_name))
        agent = agent_result.scalar_one_or_none()
        if not agent:
            return {"error": f"Agent '{agent_name}' not found."}

        fr_result = await db.execute(select(FeatureRequest).where(FeatureRequest.id == feature_id))
        feature = fr_result.scalar_one_or_none()
        if not feature:
            return {"error": "Feature request not found."}

        existing = await db.execute(
            select(FeatureVote).where(
                FeatureVote.feature_id == feature_id,
                FeatureVote.user_id == agent.id,
            )
        )
        existing_vote = existing.scalar_one_or_none()
        if existing_vote:
            existing_vote.vote = vote
        else:
            db.add(FeatureVote(feature_id=feature_id, user_id=agent.id, vote=vote))
        await db.flush()

        # Get updated vote count for broadcast
        count_result = await db.execute(
            select(func.coalesce(func.sum(FeatureVote.vote), 0)).where(
                FeatureVote.feature_id == feature_id
            )
        )
        vote_count = count_result.scalar()
        await db.commit()

    # Broadcast so frontend updates in real-time
    await broadcast_event(
        feature_update_event(
            feature_id=feature_id,
            title=feature.title,
            status=feature.status,
            vote_count=vote_count,
            update_type="voted",
        )
    )

    return {"status": "voted", "feature_id": feature_id, "vote": vote, "vote_count": vote_count}


async def agent_create_feature(agent_name: str, title: str, description: str) -> dict:
    """Create a new feature request as an agent."""
    async with async_session() as db:
        agent_result = await db.execute(select(Agent).where(Agent.agent_name == agent_name))
        agent = agent_result.scalar_one_or_none()
        if not agent:
            return {"error": f"Agent '{agent_name}' not found."}

        feature = FeatureRequest(
            id=str(uuid.uuid4()),
            title=title,
            description=description,
            status="open",
            created_by=agent.id,
            created_at=datetime.now(UTC).isoformat(),
        )
        db.add(feature)
        await db.commit()

    # Broadcast so frontend updates in real-time
    await broadcast_event(
        feature_update_event(
            feature_id=feature.id,
            title=feature.title,
            status=feature.status,
            vote_count=0,
            update_type="created",
        )
    )

    return {
        "status": "created",
        "feature_id": feature.id,
        "title": feature.title,
        "description": feature.description,
    }


async def list_all_features() -> list[dict]:
    """List all feature requests with aggregated vote counts."""
    async with async_session() as db:
        query = (
            select(
                FeatureRequest,
                func.coalesce(func.sum(FeatureVote.vote), 0).label("vote_count"),
            )
            .outerjoin(FeatureVote, FeatureRequest.id == FeatureVote.feature_id)
            .group_by(FeatureRequest.id)
            .order_by(FeatureRequest.created_at.desc())
        )
        result = await db.execute(query)
        rows = result.all()

        return [
            {
                "id": fr.id,
                "title": fr.title,
                "description": fr.description,
                "status": fr.status,
                "created_by": fr.created_by,
                "created_at": fr.created_at,
                "vote_count": int(vote_count),
            }
            for fr, vote_count in rows
        ]


async def update_agent_profile(
    agent_name: str,
    description: str | None = None,
    personality: str | None = None,
    current_task: str | None = None,
    gender: str | None = None,
) -> dict:
    """Update an agent's profile (description, personality, current task, gender)."""
    if gender is not None and gender not in ("male", "female", "non-binary"):
        return {"error": "Gender must be 'male', 'female', or 'non-binary'."}

    async with async_session() as db:
        result = await db.execute(select(Agent).where(Agent.agent_name == agent_name))
        agent = result.scalar_one_or_none()
        if not agent:
            return {"error": f"Agent '{agent_name}' not found."}

        if description is not None:
            agent.description = description
        if personality is not None:
            agent.personality = personality
        if current_task is not None:
            agent.current_task = current_task
        if gender is not None:
            agent.gender = gender

        await db.commit()

        return {
            "status": "updated",
            "agent_name": agent_name,
            "description": agent.description,
            "personality": agent.personality,
            "current_task": agent.current_task,
            "gender": agent.gender,
        }

File: server/src/services/agent-registry.ts
/**
 * Agent registration and session management.
 *
 * Single entry point: registerOrConnectAgent() handles both new registrations
 * and reconnections. Agent type is determined by the caller (MCP register tool)
 * and can be "opencode", "claude_code", or "system".
 */

import { eq, and, sql } from "drizzle-orm";
import { execSync } from "node:child_process";
import { basename, normalize } from "node:path";
import { getDb } from "../db";
import {
  agents,
  channels,
  channelMembers,
  featureRequests,
  featureVotes,
  sessions,
  users,
} from "../db/schema";
import { broadcastEvent, agentStatusEvent, channelCreatedEvent, featureUpdateEvent } from "./broadcaster";
import { generateUniqueName } from "./name-generator";
import { promptEngine } from "./prompt-engine";
import { markSessionAlive as markClaudeSessionAlive } from "../sdk/claude";

/** Derive project name from path (git repo name or folder basename) */
function deriveProjectName(projectPath: string): string {
  try {
    const result = execSync(`git -C "${projectPath}" rev-parse --show-toplevel`, {
      timeout: 5000,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    });
    return basename(result.trim());
  } catch {
    return basename(normalize(projectPath));
  }
}

function makeChannelName(projectName: string): string {
  const clean = projectName.toLowerCase().replace(/[ _]/g, "-");
  return `#project-${clean}`;
}

/**
 * Register a new agent or reconnect an existing one.
 *
 * This is the single entry point for agent login. The `sessionId` is
 * required — it's the agent's login credential.
 *
 * @param agentType - Provider type: "opencode" or "claude_code" (default: "opencode")
 */
export function registerOrConnectAgent(opts: {
  sessionId: string;
  projectPath: string;
  agentName?: string | null;
  serverUrl?: string | null;
  agentType?: string;
}): Record<string, unknown> {
  const { sessionId, projectPath, agentName, serverUrl } = opts;
  const agentType = opts.agentType ?? "opencode";

  // For Claude Code agents, mark the session as alive on registration
  if (agentType === "claude_code") {
    markClaudeSessionAlive(sessionId);
  }

  // --- Reconnect path: agent_name provided and exists ---
  if (agentName) {
    const db = getDb();
    const existing = db
      .select()
      .from(agents)
      .where(eq(agents.agentName, agentName))
      .get();

    if (existing) {
      return reconnectAgent(existing, sessionId, serverUrl ?? null, projectPath, agentType);
    }
    // Agent name provided but doesn't exist — fall through to create new
  }

  // --- New registration path ---
  return createNewAgent({
    sessionId,
    projectPath,
    agentType,
    serverUrl: serverUrl ?? null,
  });
}

function reconnectAgent(
  agent: typeof agents.$inferSelect,
  sessionId: string,
  serverUrl: string | null,
  projectPath: string,
  agentType: string = "opencode"
): Record<string, unknown> {
  const db = getDb();

  // Update invocation fields
  const updates: Partial<typeof agents.$inferInsert> = {
    providerSessionId: sessionId,
    agentType,
    status: "online",
  };
  if (serverUrl) updates.serverUrl = serverUrl;
  // Claude Code agents don't have a server URL — clear it if switching providers
  if (agentType === "claude_code" && !serverUrl) updates.serverUrl = null;
  if (projectPath) {
    updates.projectPath = projectPath;
    updates.projectName = deriveProjectName(projectPath);
  }

  db.update(agents)
    .set(updates)
    .where(eq(agents.id, agent.id))
    .run();

  // Re-fetch to get updated values
  const updated = db.select().from(agents).where(eq(agents.id, agent.id)).get()!;

  broadcastEvent(
    agentStatusEvent({
      agentName: updated.agentName,
      status: "online",
      agentType: updated.agentType,
      projectName: updated.projectName,
    })
  );

  // Look up human operator for prompt context
  const human = db.select().from(users).where(eq(users.type, "human")).get();

  const channelName = makeChannelName(updated.projectName);
  const masterPrompt = promptEngine.renderMasterPrompt({
    agentName: updated.agentName,
    agentType: updated.agentType,
    projectName: updated.projectName,
    projectChannel: channelName,
    operatorName: human?.name ?? "",
    operatorDisplayName: human?.displayName ?? "",
    operatorAbout: human?.about ?? "",
    operatorInstructions: human?.agentInstructions ?? "",
  });
  const injectPrompt = promptEngine.renderRegistrationRules({
    agentName: updated.agentName,
    projectChannel: channelName,
  });

  // Build profile reminder
  const profile: Record<string, string> = {};
  if (updated.description) profile.description = updated.description;
  if (updated.personality) profile.personality = updated.personality;
  if (updated.currentTask) profile.current_task = updated.currentTask;
  if (updated.gender) profile.gender = updated.gender;

  return {
    status: "connected",
    agent_name: updated.agentName,
    project_channel: channelName,
    master_prompt: masterPrompt,
    inject_prompt: injectPrompt,
    profile: Object.keys(profile).length > 0 ? profile : null,
  };
}

function createNewAgent(opts: {
  sessionId: string;
  projectPath: string;
  agentType: string;
  serverUrl: string | null;
}): Record<string, unknown> {
  const db = getDb();
  const now = new Date().toISOString();
  const projectName = deriveProjectName(opts.projectPath);
  const channelName = makeChannelName(projectName);

  // Generate a unique quirky name
  let agentName: string;
  let attempt = 0;
  do {
    agentName = generateUniqueName(projectName, opts.agentType, attempt);
    const collision = db
      .select()
      .from(agents)
      .where(eq(agents.agentName, agentName))
      .get();
    if (!collision) break;
    attempt++;
  } while (true);

  // Create user
  const userId = crypto.randomUUID();
  db.insert(users)
    .values({ id: userId, name: agentName, type: "agent", createdAt: now })
    .run();

  // Create agent
  db.insert(agents)
    .values({
      id: userId,
      agentName,
      agentType: opts.agentType,
      projectPath: opts.projectPath,
      projectName,
      status: "online",
      serverUrl: opts.serverUrl,
      providerSessionId: opts.sessionId,
    })
    .run();

  // Ensure project channel exists
  let channel = db.select().from(channels).where(eq(channels.name, channelName)).get();
  let createdChannel = false;
  if (!channel) {
    const channelId = crypto.randomUUID();
    db.insert(channels)
      .values({
        id: channelId,
        name: channelName,
        type: "project",
        projectPath: opts.projectPath,
        createdBy: userId,
        createdAt: now,
      })
      .run();
    channel = db.select().from(channels).where(eq(channels.id, channelId)).get()!;
    createdChannel = true;
  }

  // Add agent to project channel and #general
  db.insert(channelMembers)
    .values({ channelId: channel.id, userId, joinedAt: now })
    .run();

  const general = db.select().from(channels).where(eq(channels.name, "#general")).get();
  if (general) {
    db.insert(channelMembers)
      .values({ channelId: general.id, userId, joinedAt: now })
      .run();
  }

  // Broadcast agent online status
  broadcastEvent(
    agentStatusEvent({
      agentName,
      status: "online",
      agentType: opts.agentType,
      projectName,
    })
  );

  // Broadcast new channel creation if we created one
  if (createdChannel) {
    broadcastEvent(
      channelCreatedEvent({
        channelId: channel.id,
        channelName: channel.name,
        channelType: "project",
        projectPath: opts.projectPath,
      })
    );
  }

  // Look up human operator for prompt context
  const human = db.select().from(users).where(eq(users.type, "human")).get();

  const masterPrompt = promptEngine.renderMasterPrompt({
    agentName,
    agentType: opts.agentType,
    projectName,
    projectChannel: channelName,
    operatorName: human?.name ?? "",
    operatorDisplayName: human?.displayName ?? "",
    operatorAbout: human?.about ?? "",
    operatorInstructions: human?.agentInstructions ?? "",
  });
  const injectPrompt = promptEngine.renderRegistrationRules({
    agentName,
    projectChannel: channelName,
  });

  return {
    agent_name: agentName,
    master_prompt: masterPrompt,
    project_channel: channelName,
    inject_prompt: injectPrompt,
  };
}

/** Mark agent as offline and end active session */
export function disconnectAgent(agentName: string): Record<string, unknown> {
  const db = getDb();
  const now = new Date().toISOString();

  const agent = db.select().from(agents).where(eq(agents.agentName, agentName)).get();
  if (!agent) return { error: `Agent '${agentName}' not found.` };

  // End active sessions
  db.update(sessions)
    .set({ isActive: 0, endedAt: now })
    .where(and(eq(sessions.agentId, agent.id), eq(sessions.isActive, 1)))
    .run();

  db.update(agents).set({ status: "offline" }).where(eq(agents.id, agent.id)).run();

  broadcastEvent(
    agentStatusEvent({
      agentName,
      status: "offline",
      agentType: agent.agentType,
      projectName: agent.projectName,
    })
  );

  return { status: "disconnected", agent_name: agentName };
}

/** Update agent's last heartbeat */
export function heartbeatAgent(agentName: string): Record<string, unknown> {
  const db = getDb();
  const now = new Date().toISOString();

  const agent = db.select().from(agents).where(eq(agents.agentName, agentName)).get();
  if (!agent) return { error: `Agent '${agentName}' not found.` };

  db.update(sessions)
    .set({ lastHeartbeat: now })
    .where(and(eq(sessions.agentId, agent.id), eq(sessions.isActive, 1)))
    .run();

  return { status: "ok", agent_name: agentName };
}

/** List all agents for MCP tools */
export function listAllAgents(): Array<Record<string, unknown>> {
  const db = getDb();
  return db
    .select()
    .from(agents)
    .orderBy(agents.agentName)
    .all()
    .map((a) => {
      const entry: Record<string, unknown> = {
        name: a.agentName,
        type: a.agentType,
        project: a.projectName,
        status: a.status,
      };
      if (a.description) entry.description = a.description;
      if (a.personality) entry.personality = a.personality;
      if (a.currentTask) entry.current_task = a.currentTask;
      if (a.gender) entry.gender = a.gender;
      return entry;
    });
}

/** Update an agent's profile */
export function updateAgentProfile(
  agentName: string,
  profile: {
    description?: string | null;
    personality?: string | null;
    currentTask?: string | null;
    gender?: string | null;
  }
): Record<string, unknown> {
  if (profile.gender && !["male", "female", "non-binary"].includes(profile.gender)) {
    return { error: "Gender must be 'male', 'female', or 'non-binary'." };
  }

  const db = getDb();
  const agent = db.select().from(agents).where(eq(agents.agentName, agentName)).get();
  if (!agent) return { error: `Agent '${agentName}' not found.` };

  const updates: Partial<typeof agents.$inferInsert> = {};
  if (profile.description !== undefined) updates.description = profile.description;
  if (profile.personality !== undefined) updates.personality = profile.personality;
  if (profile.currentTask !== undefined) updates.currentTask = profile.currentTask;
  if (profile.gender !== undefined) updates.gender = profile.gender;

  db.update(agents).set(updates).where(eq(agents.id, agent.id)).run();

  const updated = db.select().from(agents).where(eq(agents.id, agent.id)).get()!;

  return {
    status: "updated",
    agent_name: agentName,
    description: updated.description,
    personality: updated.personality,
    current_task: updated.currentTask,
    gender: updated.gender,
  };
}

/** Cast a vote on a feature request as an agent */
export function agentVoteFeature(
  agentName: string,
  featureId: string,
  vote: number
): Record<string, unknown> {
  if (vote !== 1 && vote !== -1) return { error: "Vote must be +1 or -1" };

  const db = getDb();
  const agent = db.select().from(agents).where(eq(agents.agentName, agentName)).get();
  if (!agent) return { error: `Agent '${agentName}' not found.` };

  const feature = db
    .select()
    .from(featureRequests)
    .where(eq(featureRequests.id, featureId))
    .get();
  if (!feature) return { error: "Feature request not found." };

  // Upsert vote
  const existing = db
    .select()
    .from(featureVotes)
    .where(
      and(
        eq(featureVotes.featureId, featureId),
        eq(featureVotes.userId, agent.id)
      )
    )
    .get();

  if (existing) {
    db.update(featureVotes)
      .set({ vote })
      .where(
        and(
          eq(featureVotes.featureId, featureId),
          eq(featureVotes.userId, agent.id)
        )
      )
      .run();
  } else {
    db.insert(featureVotes)
      .values({ featureId, userId: agent.id, vote })
      .run();
  }

  // Get updated vote count
  const countRow = db
    .select({ total: sql<number>`coalesce(sum(${featureVotes.vote}), 0)` })
    .from(featureVotes)
    .where(eq(featureVotes.featureId, featureId))
    .get();
  const voteCount = Number(countRow?.total ?? 0);

  broadcastEvent(
    featureUpdateEvent({
      featureId,
      title: feature.title,
      status: feature.status,
      voteCount,
      updateType: "voted",
    })
  );

  return { status: "voted", feature_id: featureId, vote, vote_count: voteCount };
}

/** Create a feature request as an agent */
export function agentCreateFeature(
  agentName: string,
  title: string,
  description: string
): Record<string, unknown> {
  const db = getDb();
  const agent = db.select().from(agents).where(eq(agents.agentName, agentName)).get();
  if (!agent) return { error: `Agent '${agentName}' not found.` };

  const id = crypto.randomUUID();
  const now = new Date().toISOString();

  db.insert(featureRequests)
    .values({
      id,
      title,
      description,
      status: "open",
      createdBy: agent.id,
      createdAt: now,
    })
    .run();

  broadcastEvent(
    featureUpdateEvent({
      featureId: id,
      title,
      status: "open",
      voteCount: 0,
      updateType: "created",
    })
  );

  return {
    status: "created",
    feature_id: id,
    title,
    description,
  };
}

/** List all feature requests with vote counts */
export function listAllFeatures(): Array<Record<string, unknown>> {
  const db = getDb();
  const rows = db
    .select({
      id: featureRequests.id,
      title: featureRequests.title,
      description: featureRequests.description,
      status: featureRequests.status,
      createdBy: featureRequests.createdBy,
      createdAt: featureRequests.createdAt,
      voteCount: sql<number>`coalesce(sum(${featureVotes.vote}), 0)`,
    })
    .from(featureRequests)
    .leftJoin(featureVotes, eq(featureRequests.id, featureVotes.featureId))
    .groupBy(featureRequests.id)
    .orderBy(featureRequests.createdAt)
    .all();

  return rows.map((r) => ({
    id: r.id,
    title: r.title,
    description: r.description,
    status: r.status,
    created_by: r.createdBy,
    created_at: r.createdAt,
    vote_count: Number(r.voteCount),
  }));
}
